image-tools-interface.md

```diff
@@ -0,0 +1,52 @@
+# Image Tools Interface
+
+The below document describes an interface for the OCI Image Tools project as desired by the OCI Certification Program working group (Cert WG.)
+
+Goals:
+* An engineer can verify a container image complies with the OCI Image Spec (https://github.com/opencontainers/image-spec/releases) 
+
+Inputs:
+* (REQUIRED): Path or reference to a container image
+* (OPTIONAL): If the user would like verbose output
+* (OPTIONAL): Which version of the OCI Image Spec to validate against 
+(IF the OCI Image Tool supports multiple OCI Image Spec versions; Default to highest supported)
+
+Outputs:
+* Number of cases run
+* Number of cases passed
+* Number of cases failed
+* List of cases failed
+* Version of OCI Image Tools
+* Version of OCI Image Spec
+* (IF VERBOSE) List of cases passed
+
+## Possible Example
+
+INPUT: 
+`oci-image-tools image validate <image>`
+
+OUTPUT (default): 
+```
+OCI Image Tools vX.Y.Z validating <image> with OCI Image Spec vA.B.C:
+Case 12: FAILED: Image using reserved field key: https://github.com/opencontainers/image-spec/blob/master/descriptor.md#reserved 
+Results: 
+* Ran: 24 / 24 (100%) validation cases
+* Passed: 23 / 24 (96%)
+* Failed: 1 / 24 (4%)
+```
+
+OUTPUT (verbose):
+```
+OCI Image Tools vX.Y.Z validating <image> with OCI Image Spec vA.B.C:
```

Comment:
It might be good to look at [rust's compiler error documentation](https://doc.rust-lang.org/error-index.html) for the items that might be required here for each error. As this is, it is a little hard to parse.

Let's also take a peak at Go's output:

```
=== RUN   TestCopyDirectory
--- PASS: TestCopyDirectory (0.00s)
=== RUN   TestSimpleDiff
--- PASS: TestSimpleDiff (0.00s)
=== RUN   TestDirectoryReplace
--- PASS: TestDirectoryReplace (0.00s)
=== RUN   TestRemoveDirectoryTree
--- PASS: TestRemoveDirectoryTree (0.00s)
=== RUN   TestFileReplace
--- PASS: TestFileReplace (0.00s)
=== RUN   TestUpdateWithSameTime
--- PASS: TestUpdateWithSameTime (0.00s)
=== RUN   TestBaseDirectoryChanges
--- PASS: TestBaseDirectoryChanges (0.00s)
PASS
ok  	github.com/containerd/containerd/fs	0.007s
```

Each line is marked as "starting test" and "ending test" (`===` and `---` respectively), the we get a parseable package name at the end. Timing is less important for validation, so we don't need both, but we need a way to differentiate a result from other information.

Something like this might work:

```
* TestName1: PASSED
* TestNameFailure: FAILED https://github.com/opencontainers/image-spec/blob/master/descriptor.md#reserved
... Information about failure.
* TestName2: PASSED
# Results: 23 / 24 (96%) passed
```

Notice that the status and result lines are marked with a leading character, leaving the other output to be expressive. There may be better characters than `*` for marking these lines, but we can see how we mark what is consumable by a leading character.

Reply:
We could also use [TAP](https://en.wikipedia.org/wiki/Test_Anything_Protocol) rather than making our own output format.